gateway-ha/src/main/java/io/trino/gateway/ha/router/QueryCountBasedRouter.java

```diff
@@ -222,16 +225,16 @@ private synchronized Optional<String> getBackendForRoutingGroup(String routingGr
     }
 
     @Override
-    public String provideAdhocBackend(String user)
+    public String provideAdhocBackend(MultivaluedMap<String, String> headers)
     {
-        return getBackendForRoutingGroup("adhoc", user).orElseThrow();
+        return getBackendForRoutingGroup("adhoc", headers.getFirst(USER_HEADER)).orElseThrow();
```

Comment:
UserHeader (X-Trino-User) always has one value, so getFirst is practically the same

Reply:
> UserHeader (X-Trino-User) always has one value

No. `X-Trino-User` could be null.

> getFirst is practically the same

Yes. `user` could be null, so as `getFirst()`.